src/awkward/operations/ak_to_parquet.py

```diff
@@ -178,35 +178,114 @@ def to_parquet(
-
-    data = array
-
-    pyarrow_parquet = awkward._connect.pyarrow.import_pyarrow_parquet("ak.to_parquet")
-    fsspec = awkward._connect.pyarrow.import_fsspec("ak.to_parquet")
-
-    layout = ak.operations.ak_to_layout._impl(
-        data,
-        allow_record=True,
-        allow_unknown=False,
-        none_policy="error",
-        regulararray=True,
-        use_from_iter=True,
-        primitive_policy="error",
-        string_policy="as-characters",
-    )
-    table = ak.operations.ak_to_arrow_table._impl(
-        layout,
+    return _impl(
+        array,
+        destination,
         list_to32,
         string_to32,
         bytestring_to32,
         emptyarray_to,
         categorical_as_dictionary,
         extensionarray,
         count_nulls,
+        compression,
+        compression_level,
+        row_group_size,
+        data_page_size,
+        parquet_flavor,
+        parquet_version,
+        parquet_page_version,
+        parquet_metadata_statistics,
+        parquet_dictionary_encoding,
+        parquet_byte_stream_split,
+        parquet_coerce_timestamps,
+        parquet_old_int96_timestamps,
+        parquet_compliant_nested,  # https://issues.apache.org/jira/browse/ARROW-16348
+        parquet_extra_options,
+        storage_options,
+        write_iteratively=False,
     )
 
+
+def _impl(
+    array,
+    destination,
+    list_to32,
+    string_to32,
+    bytestring_to32,
+    emptyarray_to,
+    categorical_as_dictionary,
+    extensionarray,
+    count_nulls,
+    compression,
+    compression_level,
+    row_group_size,
+    data_page_size,
+    parquet_flavor,
+    parquet_version,
+    parquet_page_version,
+    parquet_metadata_statistics,
+    parquet_dictionary_encoding,
+    parquet_byte_stream_split,
+    parquet_coerce_timestamps,
+    parquet_old_int96_timestamps,
+    parquet_compliant_nested,  # https://issues.apache.org/jira/browse/ARROW-16348
+    parquet_extra_options,
+    storage_options,
+    write_iteratively,
+):
+    # Implementation
+    import awkward._connect.pyarrow
+
+    data = array
+
+    pyarrow_parquet = awkward._connect.pyarrow.import_pyarrow_parquet("ak.to_parquet")
+    fsspec = awkward._connect.pyarrow.import_fsspec("ak.to_parquet")
+
+    if not write_iteratively:
+        layout = ak.operations.ak_to_layout._impl(
+            data,
+            allow_record=True,
+            allow_unknown=False,
+            none_policy="error",
+            regulararray=True,
+            use_from_iter=True,
+            primitive_policy="error",
+            string_policy="as-characters",
+        )
+
+        table = ak.operations.ak_to_arrow_table._impl(
+            layout,
+            list_to32,
+            string_to32,
+            bytestring_to32,
+            emptyarray_to,
+            categorical_as_dictionary,
+            extensionarray,
+            count_nulls,
+        )
+    else:
+        layout = ak.operations.ak_to_layout._impl(
+            next(data),
```

Comment:
Python makes a distinction between an iterator (something that you can call `next` on to get another item) and an iterable (something that you can call `iter` on to get an iterator). The interface should work for iterables, so `iter(data)` will be needed at some point.